### PR TITLE
Add bounded MoE expert usage metrics to SFTTrainer

### PR DESCRIPTION
**Pull request by kaining-never-stop**
# What does this PR do?

This PR adds opt-in, evaluation-only MoE expert-usage diagnostics to `SFTTrainer`.

Setting `SFTConfig(log_expert_usage=True)` makes the trainer:

- request router logits during evaluation;
- count every Top-K expert assignment rather than only the `argmax` expert;
- exclude padding positions with `attention_mask`;
- retain counts per sample until `Accelerator.gather_for_metrics()` removes distributed-sampler tail duplicates;
- sum the gathered counts into a compact `[num_layers, num_experts]` matrix without gathering full router logits;
- log five bounded-cardinality summaries:
  - `expert_usage/normalized_entropy_mean`
  - `expert_usage/normalized_entropy_min`
  - `expert_usage/max_share_mean`
  - `expert_usage/max_share_max`
  - `expert_usage/active_fraction_mean`

The feature is disabled by default because returning router logits can increase evaluation memory use and runtime. It supports both `nll` and `chunked_nll`; Liger is rejected explicitly because that path does not expose the required standard model outputs.

Refs #4611.

## Metric cardinality

This PR intentionally does not emit one scalar key per layer and expert. The five keys above remain constant as the model grows, avoiding the roughly 6,000 time series that a 48-layer, 128-expert model would otherwise produce.

The accumulated count matrix retains the layer and expert identities needed by the targeted-LoRA use case. A public representation for that matrix can be added separately as a JSON/NumPy artifact or bounded Top-N view without turning every layer-expert pair into a time-series metric.

An earlier [fork-only implementation](https://github.com/nathon-lee/trl_woo/pull/3) explored per-layer, per-expert scalar metrics but was not submitted to `huggingface/trl`. This PR uses bounded summaries in response to the metric-cardinality concern raised in #4611.

## Counting semantics

The implementation uses `attention_mask`, so the metrics represent actual non-padding router traffic, including prompt and context tokens. A `labels != -100` mask would instead measure only loss-bearing tokens. The current choice follows the compute-throughput interpretation of expert usage.

Counts are accumulated over the full evaluation before normalization. This avoids averaging per-batch percentages, which would overweight a smaller final batch.

For distributed evaluation, counts keep a leading sample dimension until `gather_for_metrics()`. Accelerate can then remove the duplicated sampler-tail samples before the counts are summed. This is covered by a two-process case with five samples and a global batch size of four.

Padding-free batches are currently rejected, including the padding-free mode implicitly enabled by BFD/BFD-split packing. That path flattens the local batch into one row and removes `attention_mask`, so it cannot preserve the original sample dimension required for sampler-tail deduplication. Non-padding-free packing remains allowed.

Prompt Tuning and P-Tuning are supported by extending the expert-usage mask over their virtual-token embeddings. Prefix-cache PEFT methods such as Prefix Tuning and Cartridge are rejected explicitly: PEFT extends their attention mask over cached prefix positions, while the MoE router emits logits only for query-token positions.

## Validation

The focused tests cover:

- Top-K counting and padding exclusion;
- mask-shape validation;
- summary-metric values;
- default-off behavior;
- Liger compatibility validation;
- end-to-end `SFTTrainer.evaluate()` coverage for both `nll` and `chunked_nll` using an in-memory tiny Qwen3-MoE;
- an exact assertion that evaluation emits only the five bounded-cardinality keys;
- distributed-sampler tail handling.

The two-GPU validation used commit [`82aa1441`](https://github.com/huggingface/trl/commit/82aa14414d8642da5fda4dc34dc1655ba5453568), with Python 3.12.11, PyTorch 2.8.0+cu128, Transformers 5.14.1, Accelerate 1.14.0, BF16, and two NVIDIA H20 GPUs.

| Check | Result |
| --- | --- |
| Ruff on the current head | Passed; 3 changed files already formatted |
| Focused tests on the current head | 16 passed |
| Focused and validator tests in the GPU run | 10 passed |
| Qwen3-MoE, single GPU | `nll = chunked_nll = 11.931875228881836` |
| GptOss, single GPU | `nll = chunked_nll = 12.207891464233398` |
| Feature disabled | No expert-usage metrics; eval-loss delta `0` |
| Qwen3-MoE, two-process NCCL | `nll = chunked_nll = 11.748832702636719` |
| Non-divisible sampler tail | Five samples, global batch size four, observed IDs exactly `[0, 1, 2, 3, 4]` |
| Rank mapping | Rank 0 on `cuda:0`; rank 1 on `cuda:1` |
| Single/distributed summaries | All five metric deltas were `0` for both loss paths |

The single-process and two-process `eval_loss` values differ in the non-divisible-tail case because `Trainer` aggregates batch/rank scalar losses with different final-batch compositions. The cross-world loss comparison is therefore diagnostic only. Loss equivalence between `nll` and `chunked_nll` remains a hard check within each world size, while cross-world expert-usage metrics and exact sample coverage are also hard checks.

The current head, [`cf34ffd6`](https://github.com/huggingface/trl/commit/cf34ffd6dea3f1c97923d0034f22aa15a710f947), rebases the implementation onto the latest `main` and adds the explicit padding-free compatibility guard, deprecated packing-alias handling, prompt-learning mask alignment, and regression tests. The counting and aggregation path exercised by the GPU run is unchanged.

## API choices

The PR currently uses all non-padding tokens and logs bounded summaries. Feedback is welcome on follow-up APIs for:

1. exposing the full layer-by-expert matrix as an artifact or bounded Top-N representation;
2. optionally measuring loss-bearing tokens instead of all non-padding tokens.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? See #4611.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review this PR. @qgallouedec and @AmineDiro, feedback on the API choices above would be especially helpful.

---

> [!NOTE]
> **Medium Risk**
> Touches eval forward paths and MoE router-logit handling in `compute_loss`/`chunked_nll`, with explicit incompatibility guards; default-off so normal training is unchanged.
> 
> **Overview**
> Adds **`SFTConfig(log_expert_usage=True)`** (default off) so **`SFTTrainer`** can log **evaluation-only** MoE routing diagnostics without emitting per-layer/per-expert time series.
> 
> During **`evaluate()`**, the trainer requests router logits, aggregates **top-k** expert assignments per layer (padding excluded via **`attention_mask`**), gathers counts across ranks, and logs **five fixed metrics** (`eval_expert_usage/normalized_entropy_*`, `max_share_*`, `active_fraction_mean`). **`predict()`** does not collect or leak these metrics.
> 
> **`chunked_nll`** is extended via **`_get_expert_usage_counts`** / **`_summarize_expert_usage`** and patched forward output **`expert_usage_counts`**. Config validation rejects incompatible setups: Liger, padding-free / BFD packing, non-MoE models, and prefix-cache PEFT (Prefix Tuning / Cartridge); **wrapped** packing and **Prompt Tuning** remain supported. **`tests/test_sft_expert_usage.py`** covers counting, guards, and end-to-end eval behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 69d7e7b09b0c86a188893e71e730a5a68c004d56. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1753,6 +1840,30 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                    )
+                attention_mask = inputs.get("attention_mask")
+                if attention_mask is None:
+                    attention_mask = torch.ones_like(inputs["input_ids"], dtype=torch.bool)
+                expert_usage_counts = _get_expert_usage_counts(
+                    router_logits, self._num_experts_per_tok, attention_mask
+                )
+            if expert_usage_counts is None:
+                raise ValueError("Expert-usage counts were not returned by the model forward pass.")
+            expert_usage_counts = self.accelerator.gather_for_metrics(expert_usage_counts).sum(dim=0).detach()
```
[the remark this answers is not in the document]
Comment (kaining-never-stop) [reply trl/trainer/sft_trainer.py:1887]: Good catch, thanks. Padding-free collation does collapse the local batch into one row and remove `attention_mask`, so the existing per-sample `gather_for_metrics()` path cannot remove duplicated sampler-tail rows correctly. `position_ids == 0` also cannot safely recover the original rows because it marks both sequence boundaries and `pad_to_multiple_of` padding.

I kept the fix explicit and scoped in [`dc5ee0cc`](https://github.com/huggingface/trl/commit/dc5ee0cc10ba840403fba11559f69c0af0a07b8c): `log_expert_usage=True` now rejects `padding_free=True` and the BFD/BFD-split packing modes that enable padding-free collation implicitly. Non-padding-free packing remains allowed.

I also rebased onto the current `main` and added regression coverage for the explicit and implicit padding-free cases. The focused file now has `11 passed`; Ruff and formatting checks pass.

File: trl/trainer/sft_config.py [hunk 2]
```diff
@@ -307,6 +319,15 @@ class SFTConfig(_BaseConfig):
...
         super().__post_init__()
+        if self.log_expert_usage and self.use_liger_kernel:
+            raise ValueError("`log_expert_usage=True` is not currently supported with `use_liger_kernel=True`.")
+        if self.log_expert_usage and (
+            self.padding_free or (self.packing and self.packing_strategy in {"bfd", "bfd_split"})
+        ):
+            raise ValueError(
+                "`log_expert_usage=True` is not currently supported with padding-free batches, including "
+                "`packing=True` with `packing_strategy='bfd'` or `'bfd_split'`."
+            )
```
[the remark this answers is not in the document]
Comment (kaining-never-stop) [reply trl/trainer/sft_config.py]: Thanks, this ordering issue is valid. I moved the expert-usage compatibility check after `bfd-requeue` is normalized to `bfd_split`, so the guard now operates on the canonical strategy value instead of duplicating deprecated aliases.

[`d8f6ad9b`](https://github.com/huggingface/trl/commit/d8f6ad9b465cf736e741ab8346549f3739b1b6a7) adds a regression test that checks both the deprecation warning and the padding-free rejection for `bfd-requeue`. The focused suite is now `12 passed`; Ruff and formatting checks pass.

File: trl/trainer/sft_trainer.py [hunk 3]
```diff
@@ -1753,6 +1840,30 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                    raise ValueError(
+                        "The model did not return `router_logits` while `log_expert_usage=True`. "
+                        "This model family is not compatible with expert-usage logging."
+                    )
+                attention_mask = inputs.get("attention_mask")
+                if attention_mask is None:
+                    attention_mask = torch.ones_like(inputs["input_ids"], dtype=torch.bool)
+                expert_usage_counts = _get_expert_usage_counts(
+                    router_logits, self._num_experts_per_tok, attention_mask
+                )
```
[the remark this answers is not in the document]
Comment (kaining-never-stop) [reply trl/trainer/sft_trainer.py:1884]: Thanks, this is valid for embedding-based prompt-learning methods. [`cf34ffd6`](https://github.com/huggingface/trl/commit/cf34ffd6dea3f1c97923d0034f22aa15a710f947) now prepends valid virtual-token positions to the expert-usage mask for Prompt Tuning/P-Tuning on the native `nll` path; the chunked path already receives PEFT's expanded mask.

I also checked Prefix Tuning separately. In that case PEFT extends the attention mask over cached prefix positions, but the MoE model emits router logits only for query tokens, so Transformers' own load-balancing loss fails before expert-usage aggregation. The trainer now rejects prefix-cache PEFT methods such as Prefix Tuning and Cartridge explicitly instead of failing later inside the model.

The focused tests cover Prompt Tuning success and Prefix Tuning rejection for both `nll` and `chunked_nll`. The suite is now `16 passed`; Ruff and formatting checks pass.

**Comment by kaining-never-stop** [comment]
@qgallouedec, this is ready for review. I rebased it onto the current `main` and addressed Bugbot's compatibility findings through [`cf34ffd6`](https://github.com/huggingface/trl/commit/cf34ffd6dea3f1c97923d0034f22aa15a710f947) with explicit guards and regression tests for padding-free, deprecated packing aliases, and prompt-learning PEFT.

The focused suite is `16 passed`, the related padding-free collator tests are `3 passed`, and Ruff is clean. The `Tests` and `Build PR Documentation` workflows are currently `action_required` with no jobs, so they appear to need maintainer approval. Could you approve them when convenient?

@AmineDiro, since you were cc'd on #4611, feedback on the remaining API boundary would be helpful: this PR logs bounded summaries first and leaves the full layer-by-expert matrix for an artifact or bounded Top-N follow-up.

File: trl/trainer/sft_trainer.py [hunk 4]
```diff
@@ -1753,6 +1853,33 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+            if expert_usage_counts is None:
+                raise ValueError("Expert-usage counts were not returned by the model forward pass.")
+            expert_usage_counts = self.accelerator.gather_for_metrics(expert_usage_counts).sum(dim=0).detach()
+            if self._expert_usage_counts is None:
+                self._expert_usage_counts = expert_usage_counts
+            else:
+                self._expert_usage_counts += expert_usage_counts
```
[the remark this answers is not in the document]
Comment (kaining-never-stop) [reply trl/trainer/sft_trainer.py:1891]: Good catch, thanks. `predict()` and `evaluate()` share the evaluation loop, so gating collection only on `model.training` was too broad.

[`69d7e7b0`](https://github.com/huggingface/trl/commit/69d7e7b09b0c86a188893e71e730a5a68c004d56) now activates expert-usage collection only while `SFTTrainer.evaluate()` is running. The state is restored with `try/finally`, and the count buffer is cleared at the outer evaluation boundary, so `predict()` neither requests router logits for this feature nor leaves counts that a later evaluation could consume.

I added a regression test for both `nll` and `chunked_nll` that runs `predict()` followed by `evaluate()`, checks that prediction leaves no expert-usage counts, and verifies that evaluation still emits exactly the five bounded summaries. The focused suite is now `18 passed`; Ruff and formatting checks pass. This keeps the runtime behavior aligned with the eval-dataset-only scope @qgallouedec confirmed in #4611.
